### ISSUE TITLE
including synoptic and some label with eval

### ISSUE THREAD

**Issue by PhilLAL** [taurus-org/taurus#552]
Hello,
as an exercise, I would like to create a GUI including a JDraw synoptic and a label. 
This label would print a string depending on the value of a Tango attribute.

For instance, let's imagine a device sys/tg_test/1 has an attribute the short_scalar value which is regularly updated, I would like to print in the label "ODD" if this value is odd, or print "EVEN" else.

Given the following function:
```
$ cat foo.py
def oddeven(v):
    if v % 2:
        return "ODD"
    else:
        return "EVEN"
```
It seems to work:
```
$ python -c "from foo import oddeven;print oddeven(4)"
EVEN
```

I then try to launch a TaurusForm to use this function like this (in the foo.py directory):

`$ taurusform tango:sys/tg_test/1/short_scalar 'eval:@foo.*/oddeven({tango:sys/tg_test/1/short_scalar}.magnitude)'`

but I got a window with two empty fields, and 2 buttons ("Reset" and "Apply") as in this  ![screenshot](https://framapic.org/9qk59SsaYYxl/akTz09tloF6W.png)

Do you know what could miss in my code?

**Comment by cpascual** [comment]
Doing the same worked as expected in my case:

![snapshot](https://user-images.githubusercontent.com/4115224/29568063-75fc1f54-874f-11e7-8c21-62962cbf6a32.png)

Please run the taurusform command with `--taurus-log-level=trace` and post the console output. Maybe this will give us some hint....

**Comment by PhilLAL** [comment]
Here is the stdout and stderr output : 

```
MainThread     INFO     2017-08-22 15:42:15,536 TaurusRootLogger: Using "PyQt4" for Qt
MainThread     TRACE    2017-08-22 15:42:15,725 TaurusManager: reInit()
MainThread     DEBUG    2017-08-22 15:42:15,726 TaurusManager.TaurusTP: Starting TaurusTP.W001
MainThread     DEBUG    2017-08-22 15:42:15,726 TaurusManager.TaurusTP: Starting TaurusTP.W002
MainThread     DEBUG    2017-08-22 15:42:15,726 TaurusManager.TaurusTP: Starting TaurusTP.W003
MainThread     DEBUG    2017-08-22 15:42:15,726 TaurusManager.TaurusTP: Starting TaurusTP.W004
MainThread     DEBUG    2017-08-22 15:42:15,726 TaurusManager.TaurusTP: Starting TaurusTP.W005
MainThread     DEBUG    2017-08-22 15:42:15,746 TaurusManager: Found plugin TangoFactory
MainThread     DEBUG    2017-08-22 15:42:15,746 TaurusManager: Found plugin SimulationFactory
MainThread     DEBUG    2017-08-22 15:42:15,747 TaurusManager: Found plugin EvaluationFactory
MainThread     DEBUG    2017-08-22 15:42:15,748 TaurusManager: Found plugin ResourcesFactory
MainThread     WARNING  2017-08-22 15:42:15,758 TaurusForm: Cannot handle model "tango:sys/tg_test/1/short_scalar". Using default widget.
MainThread     DEBUG    2017-08-22 15:42:15,763 TaurusValueLineEdit: Exception occured while trying to attach 'tango:sys/tg_test/1/short_scalar'
MainThread     TRACE    2017-08-22 15:42:15,780 TaurusValueLineEdit: Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/taurus/qt/qtgui/base/taurusbase.py", line 742, in _attach
    self.modelObj = taurus.Manager().getObject(cls, self.modelName)
  File "/usr/local/lib/python2.7/dist-packages/taurus/core/taurusmanager.py", line 204, in getObject
    return factory.getObject(cls, name)
  File "/usr/local/lib/python2.7/dist-packages/taurus/core/taurusfactory.py", line 217, in getObject
    return self.getAttribute(name)
  File "/usr/local/lib/python2.7/dist-packages/taurus/core/tango/tangofactory.py", line 393, in getAttribute
    raise TaurusException("Invalid Tango attribute name '%s'" % attr_name)
TaurusException: Invalid Tango attribute name 'tango:sys/tg_test/1/short_scalar'


        ------------------------------------------------------------
         filename = /usr/local/lib/python2.7/dist-packages/taurus/qt/qtgui/base/taurusbase.py
         function = _attach
             line = [741]: try:
          -> line = [742]: self.modelObj = taurus.Manager().getObject(cls, self.modelName)
             line = [743]: if not self.modelObj is None:
           locals = 
                                self = <taurus.qt.qtgui.input.tauruslineedit.TaurusValueLineEdit object at 0x7f1bf8137b[...]
                                 cls = <class 'taurus.core.taurusattribute.TaurusAttribute'>
        ------------------------------------------------------------
         filename = /usr/local/lib/python2.7/dist-packages/taurus/core/taurusmanager.py
         function = getObject
             line = [203]: if factory is None: return
          -> line = [204]: return factory.getObject(cls, name)
             line = [205]: 
           locals = 
                                self = TaurusManager()
                                 cls = <class 'taurus.core.taurusattribute.TaurusAttribute'>
                             factory = TangoFactory()
                                name = tango:sys/tg_test/1/short_scalar
        ------------------------------------------------------------
         filename = /usr/local/lib/python2.7/dist-packages/taurus/core/taurusfactory.py
         function = getObject
             line = [216]: elif issubclass(cls, TaurusAttribute):
          -> line = [217]: return self.getAttribute(name)
             line = [218]: elif issubclass(cls, TaurusConfiguration):
           locals = 
                                self = TangoFactory()
                                name = tango:sys/tg_test/1/short_scalar
                                 cls = <class 'taurus.core.taurusattribute.TaurusAttribute'>
        ------------------------------------------------------------
         filename = /usr/local/lib/python2.7/dist-packages/taurus/core/tango/tangofactory.py
         function = getAttribute
             line = [392]: if params is None:
          -> line = [393]: raise TaurusException("Invalid Tango attribute name '%s'" % attr_name)
             line = [394]: 
           locals = 
                                attr = None
                                self = TangoFactory()
                           attr_name = tango:sys/tg_test/1/short_scalar
                           validator = <taurus.core.taurusvalidator.AttributeNameValidator object at 0x7f1c021b5dd0>
                              params = None
                              kwargs = {}
MainThread     WARNING  2017-08-22 15:42:15,781 TaurusForm: Cannot handle model "eval:@foo.*/oddeven({tango:sys/tg_test/1/short_scalar}.magnitude)". Using default widget.
MainThread     DEBUG    2017-08-22 15:42:15,783 TaurusValueLineEdit: Exception occured while trying to attach 'eval:@foo.*/oddeven({tango:sys/tg_test/1/short_scalar}.magnitude)'
MainThread     TRACE    2017-08-22 15:42:15,783 TaurusValueLineEdit: Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/taurus/qt/qtgui/base/taurusbase.py", line 742, in _attach
    self.modelObj = taurus.Manager().getObject(cls, self.modelName)
  File "/usr/local/lib/python2.7/dist-packages/taurus/core/taurusmanager.py", line 204, in getObject
    return factory.getObject(cls, name)
  File "/usr/local/lib/python2.7/dist-packages/taurus/core/taurusfactory.py", line 217, in getObject
    return self.getAttribute(name)
  File "/usr/local/lib/python2.7/dist-packages/taurus/core/tango/tangofactory.py", line 393, in getAttribute
    raise TaurusException("Invalid Tango attribute name '%s'" % attr_name)
TaurusException: Invalid Tango attribute name 'eval:@foo.*/oddeven({tango:sys/tg_test/1/short_scalar}.magnitude)'


        ------------------------------------------------------------
         filename = /usr/local/lib/python2.7/dist-packages/taurus/qt/qtgui/base/taurusbase.py
         function = _attach
             line = [741]: try:
          -> line = [742]: self.modelObj = taurus.Manager().getObject(cls, self.modelName)
             line = [743]: if not self.modelObj is None:
           locals = 
                                self = <taurus.qt.qtgui.input.tauruslineedit.TaurusValueLineEdit object at 0x7f1bf80a92[...]
                                 cls = <class 'taurus.core.taurusattribute.TaurusAttribute'>
        ------------------------------------------------------------
         filename = /usr/local/lib/python2.7/dist-packages/taurus/core/taurusmanager.py
         function = getObject
             line = [203]: if factory is None: return
          -> line = [204]: return factory.getObject(cls, name)
             line = [205]: 
           locals = 
                                self = TaurusManager()
                                 cls = <class 'taurus.core.taurusattribute.TaurusAttribute'>
                             factory = TangoFactory()
                                name = eval:@foo.*/oddeven({tango:sys/tg_test/1/short_scalar}.magnitude)
        ------------------------------------------------------------
         filename = /usr/local/lib/python2.7/dist-packages/taurus/core/taurusfactory.py
         function = getObject
             line = [216]: elif issubclass(cls, TaurusAttribute):
          -> line = [217]: return self.getAttribute(name)
             line = [218]: elif issubclass(cls, TaurusConfiguration):
           locals = 
                                self = TangoFactory()
                                name = eval:@foo.*/oddeven({tango:sys/tg_test/1/short_scalar}.magnitude)
                                 cls = <class 'taurus.core.taurusattribute.TaurusAttribute'>
        ------------------------------------------------------------
         filename = /usr/local/lib/python2.7/dist-packages/taurus/core/tango/tangofactory.py
         function = getAttribute
             line = [392]: if params is None:
          -> line = [393]: raise TaurusException("Invalid Tango attribute name '%s'" % attr_name)
             line = [394]: 
           locals = 
                                attr = None
                                self = TangoFactory()
                           attr_name = eval:@foo.*/oddeven({tango:sys/tg_test/1/short_scalar}.magnitude)
                           validator = <taurus.core.taurusvalidator.AttributeNameValidator object at 0x7f1c021b5dd0>
                              params = None
                              kwargs = {}
Dummy-1        TRACE    2017-08-22 15:42:18,536 TaurusManager: cleanUp()
Dummy-1        TRACE    2017-08-22 15:42:18,537 TaurusManager: [TaurusManager] cleanUp
Dummy-1        TRACE    2017-08-22 15:42:18,537 TangoFactory: [TangoFactory] cleanUp
```

**Comment by cpascual** [comment]
Your problem seems to be that the tango attribute "sys/tg_test/1/short_scalar" is not readable. 
This attribute is typically coming from the TangoTest DS. Are you sure that it is up and registered in the tango database? (refer to the Tango docs for that). 
If you want an already working environment for testing Taurus where you do not have to configure anything,  you can use the [tango-test:debian9 docker image](https://github.com/cpascual/taurus-test/tree/debian9).

But, for the purpose of your exercise, you can even do it without using Tango at all. simply use an eval attribute instead of a tango one:
```
taurusform 'eval:randint(9)' 'eval:@foo.*/oddeven({eval:randint(9)}.magnitude)'
```

**Comment by PhilLAL** [comment]
Seems strange since command
`taurusform 'eval:randint(9)' 'eval:@foo.*/oddeven({eval:randint(9)}.magnitude)'`
appears with the same empty fields.
I suspect that is some display issue with virtualbox/tangobox and not taurus. I'll check that and let you know.

**Comment by cpascual** [comment]
>  I confirm this is due to some virtualBox issue.

**Comment by cpascual** [comment]
(sorry I closed this by mistake. Already re-open)

**Comment by PhilLAL** [comment]
Thank you ;)
So now, when I launch the command
`taurusform 'eval:randint(9)' 'eval:@foo.*/oddeven({eval:randint(9)}.magnitude)'`
I get 2 labels but only the first print the random value like this: 
![screenshot](https://framapic.org/YjeQysQKnodK/m4D9zOYPn3Ay.png)

I tried also launching the command which get the TANGO attribute value:
`taurusform tango:sys/tg_test/1/short_scalar 'eval:@foo.*/oddeven({tango:sys/tg_test/1/short_scalar}.magnitude)'`
The print is similar:
![screenshot](https://framapic.org/ZqM99q1AVLCH/GMXDoV5NxRQl.png)
(Tango Access Control was misconfigured).

It seems something is wrong with the second label. Isn't it ?

**Comment by cpascual** [comment]
Ok, the relevant message is :
`MainThread     WARNING  2017-08-28 09:34:12,219 TaurusForm: Cannot handle model "eval:@foo.*/oddeven({eval:randint(9)}.magnitude)". Using default widget.`

I know I asked this before, but just in case: please make sure that the `foo` module is importable in the context in which you are launching the taurusform command (i.e., make sure that you launch the taurusform command from the directory containing `foo.py`)
In my case this works as expected:

```
$ ls
foo.py
$ cat foo.py 
def oddeven(v):
    if v % 2:
        return "ODD"
    else:
        return "EVEN"

$ python -c "from foo import oddeven;print oddeven(4)"
EVEN
$ devenv taurusform tango:sys/tg_test/1/short_scalar 'eval:@foo.*/oddeven({tango:sys/tg_test/1/short_scalar}.magnitude)'
```

...and a form just like that of https://github.com/taurus-org/taurus/issues/552#issuecomment-324028008 is shown...

**Comment by cpascual** [comment]
Also, just for debugging, the command `taurusform 'eval:@foo.*/oddeven(123)' ` should show 

![image](https://user-images.githubusercontent.com/4115224/29769691-b93c6de6-8beb-11e7-9c98-7b3d1393246e.png)

**Comment by PhilLAL** [comment]
Yes, I forgot to mention it but I launched the command into the directory where is placed foo.py
I did not define a $PYTHONPATH since I was in the same directory (when I try `PYTHONPATH="/home/gauron/taurus" && export PYTHONPATH`, it does not change anything).

When launching `taurusform 'eval:@foo.*/oddeven(123)'` I got a window with only an empty field (and probably an empty label just before).

**Comment by cpascual** [comment]
This makes no sense to me...
What version of taurus are you using?
Make sure to use the latest develop code or at least the latest release (i.e. taurus.Release.version should be at least '4.1.1')

**Comment by cpascual** [comment]
Let's try to debug more. Can you reproduce the following ipython session?

```python
In [1]: import taurus

In [2]: taurus.Release.version
Out[2]: '4.1.2-alpha'

In [3]: a=taurus.Attribute('eval:1')

In [4]: a.read().rvalue
Out[4]: <Quantity(1, 'dimensionless')>

In [5]: b=taurus.Attribute('eval:@foo.*/oddeven(123)')

In [6]: b.read().rvalue
Out[6]: 'ODD'

In [7]: ls foo.py
foo.py

In [8]: 
```

**Comment by PhilLAL** [comment]
After checking, I have a 4.0.3 installed through apt-get on a debian 9.

**Comment by cpascual** [comment]
> I have a 4.0.3

That's the problem. For this example to work, you need #423 , which was added in 4.1.

Note that there are debian packages available for the latest release: 
https://github.com/taurus-org/taurus/releases/download/4.1.1/taurus-4.1.1_debrepo.tgz

(eventually I should upload them for inclusion in the official debian repos, but now I do not have time)

**Comment by PhilLAL** [comment]
I reinstalled taurus on a new VM with pip and it works.
But I now have to solve installation of PyQt4, for which I will open another ticket.
I will come back as soon I succeed into installing a test computer for taurus.

**Comment by cpascual** [comment]
Note: if you use debian, consider installing the debian packages I mention in https://github.com/taurus-org/taurus/issues/552#issuecomment-325590007

Also, if you just want to test taurus, consider using the taurus-test docker image as I mention in https://github.com/taurus-org/taurus/issues/552#issuecomment-324235665 since it gives you taurus+tango+epics (well actually taurus is not installed, but in the README you can see one-liners for installing from git)

**Comment by PhilLAL** [comment]
> Note: if you use debian, consider installing the debian packages I mention in #552 (comment)

Ok, so for PyQt, you recommand to install with apt-get, but for all taurus package pip will be ok, that's it ?
I currently test taurus, but it will then be used as dev computer so I understand it is better to make a proper installation withouy docker. Whatever, I keep in mind there is a docker image.

**Comment by PhilLAL** [comment]
Thank you.
So I installed python-qt4 and python-qwt5-qt4 with apt-get and `python -c 'import taurus; taurus.check_dependencies()'` seems to check all dependencies.
I still havec display issues with my VM so I will check later on another computer if it works.

**Comment by cpascual** [comment]
My recommendation in general is to install as much as possible from debian packages.

In the case of taurus, if you want to test the latest stuff from the develop branch it may be better to install with pip. If you are considering doing some changes to the code yourself (e.g. for contributing patches) you could even consider installing it in ["editable" mode](https://pip.pypa.io/en/stable/reference/pip_install/#editable-installs):

```
git clone https://github.com/taurus-org/taurus.git
pip install -e ./taurus
```

**Comment by PhilLAL** [comment]
Thank you.
So I installed python-qt4 and python-qwt5-qt4 with apt-get and `python -c 'import taurus; taurus.check_dependencies()'` seems to check almost the dependencies but I did not find packages for Qub and qtcontrols:

```
$ python -c 'import taurus; taurus.check_dependencies()'
Checking required dependencies of taurus.core...
	Checking for Python >=2.6.0... [OK] (Found 2.7.13)
Checking OPTIONAL dependencies of taurus.core...
	Checking for PyTango >=7.1.0... [OK] (Found 9.2.0)
Checking required dependencies of taurus.qt...
	Checking for PyTango >=7.1.0... [OK] (Found 9.2.0)
	Checking for PyQt >=4.4.0... [OK] (Found 4.11.4)
	Checking for PyQwt >=5.2.0... [OK] (Found 5.2.3)
Checking OPTIONAL dependencies of taurus.qt...
	Checking for Qub >=1.0.0... [WARNING] (Not found])
	Checking for spyder >=3.0.0... [OK] (Found 3.1.3)
	Checking for qtcontrols >=1.0.0... [WARNING] (Not found])
```
Perhaps this will cause some issues.

I then ran 
```
In [1]: import taurus

In [2]: taurus.Release.version
Out[2]: '4.1.2-alpha'

In [3]: a=taurus.Attribute('eval:1')

In [4]: a.read().rvalue
Out[4]: <Quantity(1, 'dimensionless')>

In [5]: b=taurus.Attribute('eval:@foo.*/oddeven(123)')

In [6]: b.read().rvalue
Out[6]: 'ODD'

In [7]: ls foo.py
foo.py
```
with success.

However, I I still have display issues either with my VM and my dev server (connected with ssh -X) when launching `taurusform 'eval:@foo.*/oddeven(123)'`.
I notice that some labels are printed while the "dunamics" ones never. Could this help find any tip?

**Comment by cpascual** [comment]
> seems to check almost the dependencies but I did not find packages for Qub and qtcontrols

About the dependencies check: there is something strange going on in your installation. The result of `python -c 'import taurus; taurus.check_dependencies()'` that you posted does **not** correspond to the taurus 4.1.2-alpha version that you report in the ipython session. It most likely comes from a parallel 4.0 installation (probably the system's). Qub and qcontrols are no longer listed as dependencies in 4.1. 

Here is how it shoould look for an installation from the latest (as of now) develop commit:

```
$ python -c 'import taurus; taurus.check_dependencies()'
Dependencies for taurus 4.1.2a0:
        [*] numpy>=1.1

Extras:
Dependencies for taurus[taurus-epics]:
        [*] numpy>=1.1
        [*] pyepics>=3.2.4
Dependencies for taurus[taurus-h5file]:
        [*] numpy>=1.1
        [ ] h5file
Dependencies for taurus[taurus-qt]:
        [*] numpy>=1.1
        [*] guiqwt>=3
        [*] lxml>=2.1
        [*] ply>=2.3
        [*] qtpy>=1.2.1
        [*] PyQt4>=4.8 (not in PyPI)
        [*] PyQt4.Qwt5>=5.2 (not in PyPI)
Dependencies for taurus[taurus-qt-editor]:
        [*] numpy>=1.1
        [*] spyder>=3
Dependencies for taurus[taurus-tango]:
        [*] numpy>=1.1
        [*] PyTango>=7.1
```
 
Please check this just to avoid future problems due to these two versions coexisting.

**Comment by cpascual** [comment]
> However, I I still have display issues described in my former comment either with my VM or my dev server (connected with ssh -X) when launching taurusform 'eval:@foo.*/oddeven(123)'.
I notice that some labels are printed while the "dunamics" ones never. Could this help find any tip?

First, make sure that you are using the right version installed (as I just mentioned, it looks like you have 2 versions installed simultaneously and you may still be using 4.0 in your tests). Please, run the following commands and paste the output (and an screenshot of the GUI from the last one):

`python -c 'import taurus;print taurus.Release.version'`

`taurusform --version`

`taurusform --taurus-log-level=trace 'eval:1' 'eval:@foo.*/oddeven(123)'`

**Comment by PhilLAL** [comment]
Sorry, I am a bit sick and I made this morning wrong outputs.

So here are the steps I followed (I rechecked each one):

1. test on TangoBox VirtualBox 

First, I tested taurus on an already installed tangobox virtualbox I got from [Tango foundation website](http://www.tango-controls.org/downloads/virtual-machine/)
On this VM, I noticed display issues (see [comment](https://github.com/taurus-org/taurus/issues/552#issue-251941646)) and you explained me it was probably due to a too old Taurus version.

For this installation:
```
$ python -c 'import taurus;print taurus.Release.version'
3.6.0
```

2. test on Debian 9 distant dev server

Then, I made some tests with a distant dev server on which I installed taurus for test.
I noticed that some tests give better results (see [comment](https://github.com/taurus-org/taurus/issues/552#issuecomment-325300230)).
This computer is a development server on which was already installed Debian 9 and Tango 9
I installed taurus through aptitude.
For this installation, I got taurus version 4.0.3 (see [comment](https://github.com/taurus-org/taurus/issues/552#issuecomment-325364430)).

3. test on a Debian 9 VirtualBox

Since I wanted to modify as few as possible the dev server to ran tests without network, I installed a VirtualBox on which I can do what I want to continue testing taurus.
I installed the same configuration the dev server (Debian 9 with Tango 9).
This time, after you suggested me the version was too old, I installed taurus with pip as suggested in the doc (see [comment](https://github.com/taurus-org/taurus/issues/552#issuecomment-325656942)).
I then solved the Qt4 by installing python-qt4 and python-qwt5-qt4 through apt-get (see [comment](https://github.com/taurus-org/taurus/issues/552#issuecomment-325916952)).
In my message this morning, I made a mistake while writing my message: I copied the result from another installation (and I do not find where it is currently).
The good output is the following (I can correct it in my former message if you want):
```
$ python -c 'import taurus; taurus.check_dependencies()'
Dependencies for taurus 4.1.1:
	[*] numpy>=1.1

Extras:
Dependencies for taurus[taurus-epics]:
	[*] numpy>=1.1
	[*] pyepics>=3.2.4; extra == "taurus-epics"
Dependencies for taurus[taurus-h5file]:
	[*] numpy>=1.1
	[*] h5file; extra == "taurus-h5file"
Dependencies for taurus[taurus-qt]:
	[*] numpy>=1.1
	[*] guiqwt>=3; extra == "taurus-qt"
	[*] spyder>=3.0; extra == "taurus-qt"
	[*] qtpy>=1.2.1; extra == "taurus-qt"
	[*] lxml>=2.1; extra == "taurus-qt"
	[*] ply>=2.3; extra == "taurus-qt"
	[*] PyQt4>=4.8 (not in PyPI)
	[*] PyQt4.Qwt5>=5.2 (not in PyPI)
Dependencies for taurus[taurus-tango]:
	[*] numpy>=1.1
	[*] PyTango>=7.1; extra == "taurus-tango"
```

I ran the commands you suggested me to run, here are the outputs:
```
In [1]: import taurus

In [2]: taurus.Release.version
Out[2]: '4.1.1'

In [3]: a=taurus.Attribute('eval:1')

In [4]: a.read().rvalue
Out[4]: <Quantity(1, 'dimensionless')>

In [5]: b=taurus.Attribute('eval:@foo.*/oddeven(123)')

In [6]: b.read().rvalue
Out[6]: 'ODD'

In [7]: ls foo.py
foo.py
```
Here are the last output of the command you suggested me to run:
```
$ python -c 'import taurus;print taurus.Release.version'
4.1.1
$ taurusform --version
MainThread     INFO     2017-08-30 17:15:27,905 TaurusRootLogger: Using "PyQt4" for Qt
MainThread     INFO     2017-08-30 17:15:27,996 taurus.qt.qtgui.icon.icon: Setting Tango icon theme (from /home/tango9/.local/lib/python2.7/site-packages/taurus/qt/qtgui/icon/)
taurusform 4.1.1
$ taurusform --taurus-log-level=trace 'eval:1' 'eval:@foo.*/oddeven(123)'
MainThread     INFO     2017-08-30 17:15:35,564 TaurusRootLogger: Using "PyQt4" for Qt
MainThread     INFO     2017-08-30 17:15:35,644 taurus.qt.qtgui.icon.icon: Setting Tango icon theme (from /home/tango9/.local/lib/python2.7/site-packages/taurus/qt/qtgui/icon/)
MainThread     TRACE    2017-08-30 17:15:35,699 TaurusManager: reInit()
MainThread     DEBUG    2017-08-30 17:15:35,700 TaurusManager.TaurusTP: Starting TaurusTP.W001
MainThread     DEBUG    2017-08-30 17:15:35,700 TaurusManager.TaurusTP: Starting TaurusTP.W002
MainThread     DEBUG    2017-08-30 17:15:35,700 TaurusManager.TaurusTP: Starting TaurusTP.W003
MainThread     DEBUG    2017-08-30 17:15:35,700 TaurusManager.TaurusTP: Starting TaurusTP.W004
MainThread     DEBUG    2017-08-30 17:15:35,701 TaurusManager.TaurusTP: Starting TaurusTP.W005
MainThread     DEBUG    2017-08-30 17:15:35,701 TaurusApplication: Failed to load sardana extensions
Traceback (most recent call last):
  File "/home/tango9/.local/lib/python2.7/site-packages/taurus/qt/qtgui/application/taurusapplication.py", line 263, in __registerExtensions
    import sardana.taurus.qt.qtcore.tango.sardana
ImportError: No module named sardana.taurus.qt.qtcore.tango.sardana
MainThread     DEBUG    2017-08-30 17:15:35,705 TaurusRootLogger: cannot import epics module. Taurus will not support the "epics" scheme
MainThread     DEBUG    2017-08-30 17:15:35,705 TaurusRootLogger: cannot import epics module. Taurus will not support the "epics" scheme
MainThread     DEBUG    2017-08-30 17:15:35,705 TaurusManager: Failed to inspect taurus.core.epics
MainThread     DEBUG    2017-08-30 17:15:35,708 TaurusManager: Details:
Traceback (most recent call last):
  File "/home/tango9/.local/lib/python2.7/site-packages/taurus/core/taurusmanager.py", line 343, in _get_plugin_classes
    globals(), locals(), ['*'])
  File "/home/tango9/.local/lib/python2.7/site-packages/taurus/core/epics/__init__.py", line 70, in <module>
    from epicsfactory import *
  File "/home/tango9/.local/lib/python2.7/site-packages/taurus/core/epics/epicsfactory.py", line 34, in <module>
    import epics
ImportError: No module named epics
MainThread     DEBUG    2017-08-30 17:15:35,711 TaurusManager: Found plugin ResourcesFactory
MainThread     DEBUG    2017-08-30 17:15:35,712 TaurusManager: Found plugin TangoFactory
MainThread     DEBUG    2017-08-30 17:15:35,713 TaurusManager: Found plugin EvaluationFactory
MainThread     DEBUG    2017-08-30 17:15:35,729 TaurusPollingTimer[3000].Timer on _pollAttributes: Timer::start()
MainThread     DEBUG    2017-08-30 17:15:35,729 localhost.DefaultEvaluator.1: Transformation converted to Quantity
TimerLoop 1    DEBUG    2017-08-30 17:15:35,733 TaurusPollingTimer[3000].Timer on _pollAttributes: Timer thread starting
TimerLoop 1    DEBUG    2017-08-30 17:15:35,733 localhost.DefaultEvaluator.1: Transformation converted to Quantity
MainThread     WARNING  2017-08-30 17:15:35,738 EvaluationFactory: Problem importing "foo"
MainThread     WARNING  2017-08-30 17:15:35,740 TaurusForm: Cannot handle model "eval:@foo.*/oddeven(123)". Using default widget.
MainThread     WARNING  2017-08-30 17:15:35,742 EvaluationFactory: Problem importing "foo"
MainThread     DEBUG    2017-08-30 17:15:35,742 TaurusValue: Exception occured while trying to attach 'eval:@foo.*/oddeven(123)'
MainThread     TRACE    2017-08-30 17:15:35,766 TaurusValue: Traceback (most recent call last):
  File "/home/tango9/.local/lib/python2.7/site-packages/taurus/qt/qtgui/base/taurusbase.py", line 891, in _attach
    self.modelObj = taurus.Manager().getObject(cls, self.modelName)
  File "/home/tango9/.local/lib/python2.7/site-packages/taurus/core/taurusmanager.py", line 203, in getObject
    return factory.getObject(cls, name)
  File "/home/tango9/.local/lib/python2.7/site-packages/taurus/core/taurusfactory.py", line 293, in getObject
    return self.getAttribute(name)
  File "/home/tango9/.local/lib/python2.7/site-packages/taurus/core/evaluation/evalfactory.py", line 196, in getAttribute
    dev = self.getDevice(validator.getDeviceName(attr_name))
  File "/home/tango9/.local/lib/python2.7/site-packages/taurus/core/evaluation/evalfactory.py", line 137, in getDevice
    m = importlib.import_module(modulename)
  File "/usr/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
ImportError: No module named foo


	------------------------------------------------------------
	 filename = /home/tango9/.local/lib/python2.7/site-packages/taurus/qt/qtgui/base/taurusbase.py
	 function = _attach
	     line = [890]: try:
	  -> line = [891]: self.modelObj = taurus.Manager().getObject(cls, self.modelName)
	     line = [892]: if self.modelObj is not None:
	   locals = 
		                self = <taurus.qt.qtgui.panel.taurusvalue.TaurusValue object at 0x7fb1283ccdf8>
		                 cls = <class 'taurus.core.evaluation.evalattribute.EvaluationAttribute'>
	------------------------------------------------------------
	 filename = /home/tango9/.local/lib/python2.7/site-packages/taurus/core/taurusmanager.py
	 function = getObject
	     line = [202]: return
	  -> line = [203]: return factory.getObject(cls, name)
	     line = [204]: 
	   locals = 
		                self = TaurusManager()
		                 cls = <class 'taurus.core.evaluation.evalattribute.EvaluationAttribute'>
		             factory = EvaluationFactory()
		                name = eval:@foo.*/oddeven(123)
	------------------------------------------------------------
	 filename = /home/tango9/.local/lib/python2.7/site-packages/taurus/core/taurusfactory.py
	 function = getObject
	     line = [292]: elif issubclass(cls, TaurusAttribute):
	  -> line = [293]: return self.getAttribute(name)
	     line = [294]: # For backward compatibility
	   locals = 
		              t4_msg = The TaurusConfiguration classes are deprecated in tep14
		                self = EvaluationFactory()
		                name = eval:@foo.*/oddeven(123)
		                 cls = <class 'taurus.core.evaluation.evalattribute.EvaluationAttribute'>
	------------------------------------------------------------
	 filename = /home/tango9/.local/lib/python2.7/site-packages/taurus/core/evaluation/evalfactory.py
	 function = getAttribute
	     line = [195]: if a is None:  # if the full name is not there, create one
	  -> line = [196]: dev = self.getDevice(validator.getDeviceName(attr_name))
	     line = [197]: kwargs['storeCallback'] = self._storeAttr
	   locals = 
		                   a = None
		                self = EvaluationFactory()
		           attr_name = eval:@foo.*/oddeven(123)
		           validator = <taurus.core.evaluation.evalvalidator.EvaluationAttributeNameValidator object at[...]
		               names = ('eval://localhost/@foo.*/oddeven(123)', '@foo.*/oddeven(123)', 'oddeven(123)')
		              kwargs = {}
		            fullname = eval://localhost/@foo.*/oddeven(123)
	------------------------------------------------------------
	 filename = /home/tango9/.local/lib/python2.7/site-packages/taurus/core/evaluation/evalfactory.py
	 function = getDevice
	     line = [136]: import importlib
	  -> line = [137]: m = importlib.import_module(modulename)
	     line = [138]: except:
	   locals = 
		          modulename = foo
		           importlib = <module 'importlib' from '/usr/lib/python2.7/importlib/__init__.pyc'>
		           _safedict = {}
		                self = EvaluationFactory()
		            DevClass = <class 'taurus.core.evaluation.evaldevice.EvaluationDevice'>
		          normalname = @foo.*
		           classargs = None
		           classname = *
		            dev_name = eval://localhost/@foo.*
		             devname = foo.*
		           validator = <taurus.core.evaluation.evalvalidator.EvaluationDeviceNameValidator object at 0x[...]
		               names = ('eval://localhost/@foo.*', '@foo.*', 'foo.*')
		              groups = {'_evalclassparenths': None, '_evalclassname': '*', 'devname': '@foo.*', 'author[...]
		            fullname = eval://localhost/@foo.*
		                   d = None
	------------------------------------------------------------
	 filename = /usr/lib/python2.7/importlib/__init__.py
	 function = import_module
	     line = [36]: name = _resolve_name(name[level:], package, level)
	  -> line = [37]: __import__(name)
	     line = [38]: return sys.modules[name]
	   locals = 
		                name = foo
		             package = None
TimerLoop 1    DEBUG    2017-08-30 17:15:38,734 localhost.DefaultEvaluator.1: Transformation converted to Quantity
```
Here is the window I got: 
![screenshot](https://framapic.org/H7TnNiV4hOVO/QwSc5AkvlFEE.png)


I add also:
```
$ cat foo.py
def oddeven(v):
    if v % 2:
        return "ODD"
    else:
        return "EVEN"
$ python -c "from foo import oddeven;print oddeven(4)"
EVEN
```

**Comment by cmft** [comment]
Hi @PhilLAL ,
The error seems to be related with the `foo` module (is not in the python path).

``` 
export PYTHONPATH=/home/cfalcon/tools/test/taurus_552:$PYTHONPATH 
cd
taurusform --taurus-log-level=trace 'eval:1' 'eval:@foo.*/oddeven(123)'
``` 
or
in the path where you have `foo.py` run `taurusform --taurus-log-level=trace 'eval:1' 'eval:@foo.*/oddeven(123)'`

![taurusform1](https://user-images.githubusercontent.com/6168722/29918390-a3a2d55e-8e45-11e7-922a-46fa1a00926f.png)

**Comment by PhilLAL** [comment]
Ok, I solved the display issue. This second message mentionning PYTHONPATH made me think to a new test.
I ran all my tests from the directory where is the file `foo.py`, I did not modified PYTHONPATH because I though it was not necessary.
So I checked PYTHONPATH value and I found that it was not initialised.
I initialised it just in case: `export PYTHONPATH=/home/tango9/taurus/`
Then, launching exactly the same command `taurusform --taurus-log-level=trace 'eval:1' 'eval:@foo.*/oddeven(123)'` succeded. There was no more display issue and window was exactly as in your last message.

To be sure I did not forgot anything, I double checked by unsetting PYTHONPATH, launching again the taurusform command (it was back to the display issue), export again PYTHONPATH, and then launching taurusform command (success again).

The display issue was solved by an initialisation of PYTHONPATH.

I will now continue further to integrate a synoptic and a taurusform as just created.
